adapters/rhomarkets/src/sdk/index.ts

```diff
@@ -18,17 +29,110 @@ export type OutputDataSchemaRow = {
 };
 
 export const getUserTVLByBlock = async (blocks: BlockData) => {
-  const protocalInfos = await fetchGraphQLData(blocks.blockNumber);
-
-  const csvRows: OutputDataSchemaRow[] = protocalInfos.map((item) => ({
-    protocol: "RhoMarkets",
-    date: blocks.blockTimestamp,
-    block_number: blocks.blockNumber,
-    user_address: item.user_address,
-    market: item.market,
-    supply_token: item.supply_token,
-    borrow_token: item.borrow_token,
-  }));
+  const protocalInfos = await fetchGraphQLData(
+    blocks.blockNumber,
+    CHAINS.SCROLL,
+    PROTOCOLS.RHO_MARKETS
+  );
+
+  const marketInfos = await getMarketInfos(
+    "0x8a67AB98A291d1AEA2E1eB0a79ae4ab7f2D76041"
+  );
+
+  const newStates: any[] = protocalInfos.filter(
+    (x) =>
+      (x.borrow_token > 0 || x.supply_token > 0) && x.market && x.user_address
+  );
+
+  const marketsMapping: any = {};
+  for (let marketInfo of marketInfos) {
+    marketsMapping[marketInfo.address] = {
+      address: marketInfo.address,
+      exchangeRate: marketInfo.exchangeRateStored,
+      decimals: marketInfo.decimals,
+    };
+  }
+
+  const publicClient = createPublicClient({
+    chain: extractChain({ chains: [scroll], id: CHAINS.SCROLL }),
+    transport: http(RPC_URLS[CHAINS.SCROLL]),
+  });
+
+  console.log(`Will update all borrow balances for ${newStates.length} states`);
+  for (var i = 0; i < newStates.length; i += 1000) {
+    const start = i;
+    const end = i + 1000;
+    var subStates = newStates.slice(start, end);
+    console.log(`Updating borrow balances for ${start} - ${end}`);
+
+    let borrowBalanceResults = await publicClient.multicall({
+      contracts: subStates
+        .map((m) => {
+          return [
+            {
+              address: m.market,
+              abi: ltokenAbi,
+              functionName: "borrowBalanceStored",
+              args: [m.user_address],
+            },
+          ];
+        })
+        .flat() as any,
+      blockNumber: BigInt(blocks.blockNumber),
+    });
+
+    let supplyBalanceResults = await publicClient.multicall({
+      contracts: subStates
+        .map((m) => {
+          return [
+            {
+              address: m.market,
+              abi: ltokenAbi,
+              functionName: "balanceOf",
+              args: [m.user_address],
+            },
+          ];
+        })
+        .flat() as any,
+      blockNumber: BigInt(blocks.blockNumber),
+    });
+
+    for (var j = 0; j < subStates.length; j++) {
+      subStates[j].borrow_token = formatUnits(
+        (borrowBalanceResults[j]?.result as bigint) || 0n,
+        subStates[j].decimals
+      );
+
+      subStates[j].supply_token = formatUnits(
+        (((supplyBalanceResults[j]?.result as bigint) || 0n) *
+          marketsMapping[subStates[j].market].exchangeRate) /
+          10n ** 18n,
```

Comment:
https://docs.compound.finance/v2/ctokens/#exchange-rate

the exchange rate may not be just 18 decimals according to compound docs

Reply:
I think it fine, because our excahneg is fixed 18 decimals

Reply:
ExhangeRate not the same as borrowBalanceStored decimals